### PR TITLE
boards/lora-e5-dev: initial support

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR adds support for [LoRa-E5 Development Board](https://www.google.com/search?channel=fs&client=ubuntu&q=seedstudio+lora-e5). 

Missing form this PR:
- [ ] support for the integrated temperature sensor
- [ ] support for the integrated spi-flash (no documentation available AFAIKT)

There seems to be an issue with I2C but I experienced the same issue with `nucleo-wl55jc`, so I think its unrelated to this PR and I'm confident on the pin mapping, so I kept the support. Maybe @MrKevinWeiss can use HIL to verify that `stm32wl` i2c is not broken?

### Testing procedure

Tested main functionalities:

- spi ( bme280_spi)

```
2021-07-20 13:46:27,385 # main(): This is RIOT! (Version: 2021.10-devel-170-g05efb-wip/lora_e5)
2021-07-20 13:46:27,386 # BMX280 test application
2021-07-20 13:46:27,386 #
2021-07-20 13:46:27,393 # +------------Initializing------------+
2021-07-20 13:46:27,393 # Initialization successful
2021-07-20 13:46:27,394 #
2021-07-20 13:46:27,398 # +------------Calibration Data------------+
2021-07-20 13:46:27,399 # dig_T1: 28497
2021-07-20 13:46:27,399 # dig_T2: 26274
2021-07-20 13:46:27,400 # dig_T3: 50
2021-07-20 13:46:27,404 # dig_P1: 37174
2021-07-20 13:46:27,404 # dig_P2: -10773
2021-07-20 13:46:27,405 # dig_P3: 3024
2021-07-20 13:46:27,405 # dig_P4: 9121
2021-07-20 13:46:27,406 # dig_P5: -67
2021-07-20 13:46:27,410 # dig_P6: -7
2021-07-20 13:46:27,410 # dig_P7: 12300
2021-07-20 13:46:27,410 # dig_P8: -12000
2021-07-20 13:46:27,411 # dig_P9: 5000
2021-07-20 13:46:27,411 # dig_H1: 75
2021-07-20 13:46:27,415 # dig_H2: 334
2021-07-20 13:46:27,415 # dig_H3: 0
2021-07-20 13:46:27,416 # dig_H4: 393
2021-07-20 13:46:27,416 # dig_H5: 50
2021-07-20 13:46:27,416 # dig_H6: 30
2021-07-20 13:46:27,416 #
2021-07-20 13:46:27,421 # +--------Starting Measurements--------+
2021-07-20 13:46:27,424 # [bmx280] _do_measurement: measurement data ready
2021-07-20 13:46:27,439 # [bmx280] _do_measurement: measurement data ready
2021-07-20 13:46:27,445 # [bmx280] _do_measurement: measurement data ready
2021-07-20 13:46:27,445 # Temperature [°C]: 26.14
2021-07-20 13:46:27,450 #    Pressure [Pa]: 99053
2021-07-20 13:46:27,451 #   Humidity [%rH]: 52.67
2021-07-20 13:46:27,451 #
2021-07-20 13:46:27,454 # +-------------------------------------+
2021-07-20 13:46:27,454 #
2021-07-20 13:46:29,459 # [bmx280] _do_measurement: measurement data ready
2021-07-20 13:46:29,467 # [bmx280] _do_measurement: measurement data ready
2021-07-20 13:46:29,475 # [bmx280] _do_measurement: measurement data ready
2021-07-20 13:46:29,480 # Temperature [°C]: 26.16
2021-07-20 13:46:29,480 #    Pressure [Pa]: 99076
2021-07-20 13:46:29,481 #   Humidity [%rH]: 52.62
2021-07-20 13:46:29,481 #
2021-07-20 13:46:29,485 # +-------------------------------------+
```

- rtt
```
2021-07-20 09:11:05,949 # main(): This is RIOT! (Version: 2021.10-devel-170-g43cad-wip/lora_e5)
2021-07-20 09:11:05,949 #
2021-07-20 09:11:05,951 # RIOT RTT low-level driver test
2021-07-20 09:11:05,952 # RTT configuration:
2021-07-20 09:11:05,955 # RTT_MAX_VALUE: 0x0000ffff
2021-07-20 09:11:05,956 # RTT_FREQUENCY: 1024
2021-07-20 09:11:05,956 #
2021-07-20 09:11:05,960 # Testing the tick conversion
2021-07-20 09:11:05,961 # Trying to convert 1 to seconds and back
2021-07-20 09:11:05,965 # Trying to convert 256 to seconds and back
2021-07-20 09:11:05,971 # Trying to convert 65536 to seconds and back
2021-07-20 09:11:05,971 # Trying to convert 16777216 to seconds and back
2021-07-20 09:11:05,978 # Trying to convert 2147483648 to seconds and back
2021-07-20 09:11:05,978 # All ok
2021-07-20 09:11:05,978 #
2021-07-20 09:11:05,979 # Initializing the RTT driver
2021-07-20 09:11:06,139 # This test will now display 'Hello' every 5 seconds
2021-07-20 09:11:06,140 #
2021-07-20 09:11:06,140 # RTT now: 1
2021-07-20 09:11:06,144 # Setting initial alarm to now + 5 s (5121)
2021-07-20 09:11:06,150 # rtt_get_alarm() PASSED
2021-07-20 09:11:06,151 # Done setting up the RTT, wait for many Hellos
2021-07-20 09:11:11,141 # Hello
2021-07-20 09:11:16,140 # Hello
2021-07-20 09:11:21,139 # Hello
2021-07-20 09:11:26,139 # Hello
2021-07-20 09:11:31,138 # Hello
```
- loramac

```
2021-07-20 10:13:08,853 # All up, running the shell now
: loramac set deveui 00D23F63A8D87785
2021-07-20 10:13:12,822 # loramac set deveui 00D23F63A8D87785
: loramac set appeui 70B3D57ED00204EC
2021-07-20 10:13:15,817 # loramac set appeui 70B3D57ED00204EC
: loramac set appkey 916145C2F84B9B30584164B8625B50EE
2021-07-20 10:13:18,962 # loramac set appkey 916145C2F84B9B30584164B8625B50EE
: loramac join otaa
2021-07-20 10:13:20,813 # loramac join otaa
2021-07-20 10:13:29,141 # Join procedure succeeded!
loramac tx "hi"
2021-07-20 10:13:42,814 # loramac tx "hi"
2021-07-20 10:13:50,162 # Received ACK from network
2021-07-20 10:13:50,163 # Message sent with success
```

### CODE REVIEW

**Comment by benpicco** [comment]
> - [ ] support for the integrated spi-flash (no documentation available AFAIKT)

You could make some educated guesses, e.g. 256 byte pages and 4k sectors, so

```C
static mtd_spi_nor_t lora_e5_nor_dev = {
    .base = {
        .driver = &mtd_spi_nor_driver,
        .page_size = 256,
        .pages_per_sector = 16,
    },
    .params = &lora_e5_nor_params,
};
```

The number of pages should be auto-detected if not given and the timings are minimum values anyway.
You can use `tests/mtd_raw` to verify that the flash was properly identified.

File: boards/lora-e5-dev/Makefile.dep [hunk 1]
```diff
@@ -0,0 +1,11 @@
+ifneq (,$(filter stdio_uart,$(USEMODULE)))
+  FEATURES_REQUIRED += periph_uart
+endif
```
Comment (benpicco) [inline boards/lora-e5-dev/Makefile.dep:3]: isn't that already the dependency of `stdio_uart`?

But should we have

```suggestion
ifneq (,$(filter netdev_default,$(USEMODULE)))
  USEMODULE += sx126x…
endif
```
?

File: boards/lora-e5-dev/board.c [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+#if IS_USED(MODULE_SX126X_STM32WL)
+/**
+ * @brief Callback to set RF switch mode
+ *
+ * This function sets the GPIO's wired to the SP3T RF Switch. LoRa-E5-dev
+ * supports two modes of operation.
+ */
+void lora_e5_dev_sx126x_set_rf_mode (sx126x_t *dev, sx126x_rf_mode_t rf_mode)
```
Comment (benpicco) [inline boards/lora-e5-dev/board.c:53]:
```suggestion
void lora_e5_dev_sx126x_set_rf_mode(sx126x_t *dev, sx126x_rf_mode_t rf_mode)
```

File: boards/lora-e5-dev/include/board.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * @name    Sub-GHz radio (LoRa) configuration
+ * @{
+ */
+#define SX126X_PARAM_SPI                    (SPI_DEV(0))
+#if IS_USED(MODULE_SX126X_STM32WL)
+extern void lora_e5_dev_sx126x_set_rf_mode(sx126x_t *dev, sx126x_rf_mode_t rf_mode);
+#define SX126X_PARAM_SET_RF_MODE_CB         lora_e5_dev_sx126x_set_rf_mode
+#define SX126X_PARAM_TYPE                   SX126X_TYPE_STM32WL
+#endif
```
Comment (benpicco) [inline boards/lora-e5-dev/include/board.h:41]: If the radio module is part of the CPU the config should probably be moved there as it's not board specific.

File: boards/lora-e5-dev/include/periph_conf.h [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        .bus        = APB12,
+        .irqn       = LPUART1_IRQn,
+        .type       = STM32_LPUART,
+        .clk_src    = 0, /* Use APB clock */
+    },
+};
+
+#define UART_0_ISR          (isr_usart1)
+#define UART_1_ISR          (isr_usart2)
+#define UART_2_ISR          (isr_lpuart1)
```
Comment (benpicco) [inline boards/lora-e5-dev/include/periph_conf.h:89]: I'm surprised this works with the parentheses given that it's used as a function declaration.

File: boards/lora-e5-dev/include/periph_conf.h [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+ */
+static const spi_conf_t spi_config[] = {
+    {
+        .dev      = SUBGHZSPI, /* Internally connected to Sub-GHz radio Modem  */
+        .rccmask  = RCC_APB3ENR_SUBGHZSPIEN,
+        .apbbus   = APB3,
+    }
+/* SUBGHZ DEBUG PINS use the SPI1 pins */
+#if !IS_ACTIVE(CONFIG_STM32_WLX5XX)
+    ,{
```
Comment (benpicco) [inline boards/lora-e5-dev/include/periph_conf.h:105]:
```suggestion
    },
/* SUBGHZ DEBUG PINS use the SPI1 pins */
#if !IS_ACTIVE(CONFIG_STM32_WLX5XX)
    {
```

File: boards/nucleo-wl55jc/Makefile.dep [hunk 6]
```diff
@@ -1,7 +1,7 @@
 ifneq (,$(filter stdio_uart,$(USEMODULE)))
   FEATURES_REQUIRED += periph_lpuart
 endif
-ifneq (,$(filter netdev_default,$(USEMODULE)))
+ifneq (,$(filter sx126x%,$(USEMODULE)))
```
Comment (benpicco) [inline boards/nucleo-wl55jc/Makefile.dep:4]: why change this?

Comment (fjmolinas) [reply boards/nucleo-wl55jc/Makefile.dep:4]: I think it a mishap of rebasing my branch that was based on an earlier version of the branch adding support for the `sx126x` spi radio.

File: boards/lora-e5-dev/include/periph_conf.h (hunk 4 above)
Comment (fjmolinas) [reply boards/lora-e5-dev/include/periph_conf.h:89]: Hmm this is a copy pasta... but seems to be the case in many stm32 BOARDs, I'll fix here at least

File: boards/lora-e5-dev/include/board.h (hunk 3 above)
Comment (fjmolinas) [reply boards/lora-e5-dev/include/board.h:41]: Hmmm I'm not sure if its CPU specific, the lora-e5 does not have the same RF modes as the nucleo-wl55jc, I'm not sure
if this is BOARD or CPU_MODEL specific.

**Comment by fjmolinas** [comment]
While running `compile_and_test_for_board` I realized that this is actually not a dual-core CPU, so I adjusted some of the flash page code accordingly...

> > * [ ]  support for the integrated spi-flash (no documentation available AFAIKT)
> 
> You could make some educated guesses, e.g. 256 byte pages and 4k sectors, so
> 
> ```c
> static mtd_spi_nor_t lora_e5_nor_dev = {
>     .base = {
>         .driver = &mtd_spi_nor_driver,
>         .page_size = 256,
>         .pages_per_sector = 16,
>     },
>     .params = &lora_e5_nor_params,
> };
> ```
> 
> The number of pages should be auto-detected if not given and the timings are minimum values anyway.
> You can use `tests/mtd_raw` to verify that the flash was properly identified.

I tried that and it didn't work, I'll leave at is for now.

**Comment by fjmolinas** [comment]
Two failing builds in murdock, addressed in last commits, may I squash @benpicco?

**Comment by benpicco** [comment]
sure please squash

**Comment by fjmolinas** [comment]
> sure please squash

Squashed, I split the commits regarding CPU changes.

**Review by benpicco — APPROVED** [review]

**Comment by fjmolinas** [comment]
Thanks for the quick review!

**Pull request merged**